Mathlib/RingTheory/OreLocalization/Algebra.lean

```diff
@@ -0,0 +1,210 @@
+
+end One
+
+section CommMonoid
+
+variable {R A : Type*} [CommMonoid R] [CommMonoid A] [MulAction R A] [IsScalarTower R A A]
+  {S : Submonoid R}
+
+instance commMonoid' : CommMonoid (A[S⁻¹]) where
+  mul_assoc a b c := by
+    induction' a with a
+    induction' b with b
+    induction' c with c
+    simp only [OreLocalization.mul_def, mul_assoc]
+  one_mul a := by
+    induction' a with a
+    simp only [OreLocalization.one_def', OreLocalization.mul_def, one_mul, mul_one]
+  mul_one a := by
+    induction' a with a s
+    simp only [OreLocalization.one_def', OreLocalization.mul_def, mul_one, one_mul]
+  mul_comm a b := by
+    induction' a with a
+    induction' b with b
+    simp only [OreLocalization.mul_def, mul_comm]
+
+end CommMonoid
+
+section CommSemiring
+
+variable {R A : Type*} [CommMonoid R] [CommSemiring A] [DistribMulAction R A] [IsScalarTower R A A]
+  {S : Submonoid R}
+
+theorem left_distrib' (a b c : A[S⁻¹]) :
+    a * (b + c) = a * b + a * c := by
+  induction' a with a r
+  induction' b with b s
+  induction' c with c t
+  rcases oreDivAddChar' b c s t with ⟨r₁, s₁, h₁, q⟩; rw [q]; clear q
+  simp only [OreLocalization.mul_def]
+  rcases oreDivAddChar' (a * b) (a * c) (s * r) (t * r) with ⟨r₂, s₂, h₂, q⟩; rw [q]; clear q
+  rw [oreDiv_eq_iff]
+  obtain ⟨t, ht⟩ := t
+  obtain ⟨s₂, hs₂⟩ := s₂
+  simp only [Submonoid.mk_smul, Submonoid.coe_mul] at h₁ h₂ ⊢
+  rw [h₂, h₁]
+  use s₁ * s * r, s₂ * s * r
+  obtain ⟨s₁, hs₁⟩ := s₁
+  obtain ⟨s, hs⟩ := s
+  obtain ⟨r, hr⟩ := r
+  simp only [Submonoid.mk_mul_mk, smul_add, Submonoid.mk_smul] at h₁ h₂ ⊢
+  constructor
+  · rw [mul_add, smul_add]
+    congr 1
+    · rw [smul_smul, mul_comm a (s₁ • b), show s₁ * s * r * s₂ = (s₂ * s * r) * s₁ by
+        rw [mul_comm, mul_assoc, mul_assoc, mul_assoc, mul_comm s₁, mul_assoc]]
+      rw [mul_smul]
+      congr 1
+      rw [mul_comm, smul_mul_assoc]
+    · rw [h₁, mul_assoc s₂, h₂, mul_comm r₂, mul_assoc, mul_comm r₁, mul_smul, mul_smul (t * r)]
+      congr 1
+      rw [smul_smul, mul_comm, mul_smul, mul_comm a, mul_comm a, smul_mul_assoc]
+  · rw [mul_assoc s₂, h₂, h₁]
+    repeat rw [← mul_assoc]
+    congr 2
+    rw [mul_right_comm _ r, mul_right_comm _ r]
+    congr 1
+    rw [mul_comm, mul_assoc, mul_comm r₁]
+
+instance instCommSemiring' : CommSemiring A[S⁻¹] where
+  __ := commMonoid'
+  left_distrib := left_distrib'
+  right_distrib a b c := by
+    rw [mul_comm, mul_comm a, mul_comm b, left_distrib']
+  zero_mul a := by
+    induction' a with a s
+    rw [OreLocalization.zero_def, OreLocalization.mul_def, zero_mul,
+      mul_one, zero_oreDiv, zero_oreDiv]
+  mul_zero a := by
+    induction' a with a s
+    rw [OreLocalization.zero_def, OreLocalization.mul_def, mul_zero,
+      one_mul, zero_oreDiv, zero_oreDiv]
+
+end CommSemiring
+
+section Algebra
+
+variable (R A : Type*) [cR : CommSemiring R] [cA : CommSemiring A] [cRA : Algebra R A]
+    (S : Submonoid R)
+
+def instAlgebra' : Algebra (R[S⁻¹]) (A[S⁻¹]) where
```

Comment:
while defs shouldn't

Reply:
It was an instance until Eric pointed out diamond possibilities and then I found one. It's the example at the bottom of the file. Can I make this defeq somehow?